modules/core/CMakeLists.txt

```diff
@@ -143,6 +143,10 @@ endif()
 
 ocv_create_module(${extra_libs})
 
+if(BUILD_ZLIB)
+  add_dependencies(${the_module} ${ZLIB_LIBRARY})
+endif()
```

Comment:
hey sorry i didn't see this comment when it happened and only saw now that it's been closed 🤦🏽‍♂️

`add_dependencies` is required since target_link_libraries does not add a build order dependency if the library is not built by the build system. from [target_link_libraries](https://cmake.org/cmake/help/latest/command/target_link_libraries.html):

> The named target must be created by [add_library()](https://cmake.org/cmake/help/latest/command/add_library.html#command:add_library) within the project or as an [IMPORTED library](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#imported-targets). If it is created within the project an ordering dependency will automatically be added in the build system to make sure the named library target is up-to-date before the <target> links.

Also [add_dependencies](https://cmake.org/cmake/help/latest/command/add_dependencies.html?highlight=add_dependencies) not add a duplicate linkage to zlib. it tells the build system that this is a dependency of the target and must be built before it.

i tested with and without this and it only worked with the `add_dependency` so i believe this is correct and necessary.